rs/system_api/src/lib.rs

```diff
@@ -2224,17 +2234,40 @@ impl SystemApi for SystemApiImpl {
     }
 
     fn ic0_call_cycles_add(&mut self, amount: u64) -> HypervisorResult<()> {
-        let result = self.ic0_call_cycles_add_helper("ic0_call_cycles_add", Cycles::from(amount));
+        let result = self
+            .ic0_call_cycles_add_helper(
+                "ic0_call_cycles_add",
+                CyclesAmountType::Exact(Cycles::from(amount)),
+            )
+            .map(|_| ());
         trace_syscall!(self, CallCyclesAdd, result, amount);
         result
     }
 
     fn ic0_call_cycles_add128(&mut self, amount: Cycles) -> HypervisorResult<()> {
-        let result = self.ic0_call_cycles_add_helper("ic0_call_cycles_add128", amount);
+        let result = self
+            .ic0_call_cycles_add_helper("ic0_call_cycles_add128", CyclesAmountType::Exact(amount))
+            .map(|_| ());
         trace_syscall!(self, CallCyclesAdd128, result, amount);
         result
     }
 
+    fn ic0_call_cycles_add128_up_to(
+        &mut self,
+        amount: Cycles,
+        dst: usize,
+        heap: &mut [u8],
+    ) -> HypervisorResult<()> {
+        let result = self.ic0_call_cycles_add_helper(
+            "ic0_call_cycles_add128_up_to",
+            CyclesAmountType::UpTo(amount),
+        );
+        trace_syscall!(self, CallCyclesAdd128UpTo, result, amount);
```

Comment:
Should we trace the syscall if it fails? (note that we only do `result?` on the next line)